File: Stage06-FromFile/C#/Player.cs
using System;
using System.Collections.Generic;

namespace Adventure_06_Improvements
{
    internal static class Player
    {
        public static string Name = "";
        public static int Health = 100;
        public static int Strength = 100;
        public static string Character = "";
        public static string ItemInHand = "";
        public static List<string> Characters = new List<string>();
        public static List<string> Inventory = new List<string>();

        public static void AddToInventory(string item)
        {
            /// add an item to player inventory ///
            if (!Inventory.Contains(item))
                Inventory.Add(item);
        }
        public static string Attack(Location here)
        {
            string enemy = here.Enemy;
            string message = $"You attack the {enemy}";
            int damage = 10;
            string useItem = "fist";
            foreach(string item in Inventory)
            {
                if(Shared.Items[item] is Weapon)
                {
                    Weapon weapon = (Weapon)Shared.Items[item];
                    damage = weapon.Damage;
                    useItem = $" the {item}";
                    break;
                }
            }
            message += $" with {useItem} inflicting {damage} damage points";
            Shared.Enemies[enemy].ReceiveAttack(damage);

            return message;
        }
        public static void DisplayInventory()
        {
            /// display player's inventory ///
            Console.WriteLine(new string('═', Console.WindowWidth - 1));
            Console.WriteLine("Player properties:");
            Console.WriteLine(new string('═', Console.WindowWidth - 1));
            Console.Write("Characters available: ");
            foreach(string character in Characters)
                Console.Write($"{character}, ");
            Console.WriteLine($"\n{new string('═', Console.WindowWidth - 1)}");
            Console.WriteLine($"Name:                 {Name}");
            Console.WriteLine($"Health:               {Health}");
            Console.WriteLine($"Strength:             {Strength}");
            Console.WriteLine($"Character:            {Character}");
            Console.Write($"Inventory:            ");
            foreach (string item in Inventory)
                Console.Write($"{item}, ");
            Console.WriteLine(new string('═', Console.WindowWidth - 1));
        }
        public static void DisplayPlayer()
        {
            /// main use for debug. prints all player properties ///
            Console.WriteLine(new string('═', Console.WindowWidth - 1));
            Console.WriteLine("Player properties:");
            Console.WriteLine(new string('═', Console.WindowWidth - 1));
            Console.Write("Characters available: ");
            foreach (string character in Characters)
                Console.Write($"{character}, ");
            Console.WriteLine($"\n{new string('═', Console.WindowWidth - 1)}");
            Console.WriteLine($"Name:                 {Name}");
            Console.WriteLine($"Health:               {Health}");
            Console.WriteLine($"Strength:             {Strength}");
            Console.WriteLine($"Character:            {Character}");
            Console.Write($"Inventory:            ");
            foreach (string item in Inventory)
                Console.Write($"{item}, ");
            Console.WriteLine($"\n{new string('═', Console.WindowWidth - 1)}");
        }
        public static string GetProperty(string propertyName)
        {
            ///  used by narrator class to get a player properties ///
            if (propertyName.Contains("name"))
                return Name;
            else if (propertyName.Contains("health"))
                return Health.ToString();
            else if (propertyName.Contains("strength"))
                return Strength.ToString();
            else if (propertyName.Contains("character"))
                return Character;
            return "";
        }
        public static void ReceiveAttack(int damage)
        {
            Health -= damage;
            if(Health < 0)
                Health = 0;
        }

        public static void RemoveFromInventory(string item)
        {
            ///  remove an item from player inventory ///
            if(Player.Inventory.Contains(item))
                Player.Inventory.Remove(item);
        }
        public static void UpdateStats(int characterIndex)
        {
            ///  modify health and strength depending on character selected ///
            Health += characterIndex * 2;
            Strength += characterIndex * 2;
        }
    }
}

File: Stage06-FromFile/Python/player.py
''' This is a Python Module equivalent to a static class '''
import shared, weapon
# properties: Can be directly accessed for read/write

name:str = ""
health:int = 100
strength:int = 100
character:str = ""
characters:list = []
inventory:list = []

def add_to_inventory(item:str) -> None:
	''' add an item to player inventory '''
	global inventory
	if item not in inventory:
		inventory.append(item)

def attack(here:object):
	message = f"You attack the {here.enemy} "
	damage = 10
	use_item = "your fist"
	for item in inventory:
		if isinstance(shared.items[item], weapon.Weapon):
			damage = shared.items[item].damage
			use_item = f" the {item}"
			break
	
	message += f"with {use_item} inflicting {damage} damage points"
			
	shared.enemies[here.enemy].receive_attack(damage)
	
	return message	
		
def display_inventory() -> None:
	''' display player's inventory '''
	# global not needed as the value of inventory is not being changed
	if len(inventory) == 0:
		print("Your inventory is empty")
	else:
		print("In your inventory you have:")
		output:str = ""
		for item in inventory:
			output += f"{item}, "
		output = output.strip()[:-1]
		print(output)
		
def display_player() -> None:
	''' main use for debug. prints all player properties '''
	print(''.ljust(80,'─'))
	print("Player properties:")
	print(''.ljust(80,'─'))
	print(f"Characters available: {characters}")
	print(''.ljust(80,'─'))
	print(f"Name:                 {name}")
	print(f"Health:               {health}")
	print(f"Strength:             {strength}")
	print(f"Character:            {character}")
	print(f"Inventory:            {inventory}")
	print(''.ljust(80,'─'))

def get_property(property_name):
	''' used by narrator class to get a player property '''
	if "name" in property_name:
		return name
	elif "health" in property_name:
		return health
	elif "strength" in property_name:
		return strength
	elif "character" in property_name:
		return character
 
def receive_attack(damage):
	global health
	health -= damage
	if health <= 0:
		health = 0	

def remove_from_inventory(item:str) -> None:
	''' remove an item to player inventory '''
	global inventory
	if item in inventory:
		inventory.remove(item)
  
def update_stats(character_index:int) -> None:
	''' modify health and strength depending on character selected '''
	global health, strength # major failing in Python. Declared in body of script = global?
	health += character_index * 2
	strength -= character_index * 2